src/dblib_stubs.c

```diff
@@ -89,7 +127,10 @@ static int msg_handler(DBPROCESS *dbproc, DBINT msgno, int msgstate, int severit
 
   caml_callback3(*handler, vseverity, vline, vmsg);
 
-  CAMLreturn(INT_CANCEL); /* should not return */
+  if (had_lock) {
+    release_lock(dbproc);
+  }
+  CAMLreturn(0);
```

Comment:
Note: This used to have a comment `/* should not return */`, but this function will return as long as the callback doesn't throw an exception. Also the Sybase docs for `dbmsghandle` say:

> The message handler must return a value of 0 to DB-Library.

http://infocenter.sybase.com/help/index.jsp?topic=/com.sybase.help.ocs_12.5.1.dblib/html/dblib/X12458.htm

Reply:
Oh, and it's not entirely clear to me if it's safe to call CAMLreturn without the lock, but i don't know how to trigger CAMLreturn's cleanups otherwise?